EndlessClient/Input/FunctionKeyHandler.cs

```diff
@@ -27,12 +31,17 @@ protected override Option<Keys> HandleInput()
                 if (IsKeyHeld(key))
                 {
                     var isShiftHeld = IsKeyHeld(Keys.LeftShift) || IsKeyHeld(Keys.RightShift);
-                    if (_functionKeyController.SelectSpell(key - Keys.F1, isShiftHeld))
+                    var isFrozen = _playerInfoProvider.IsPlayerFrozen;
+
+                    if (isFrozen)
+                        return Option.None<Keys>();
```

Comment:
You need to decide where the frozen logic is going to be handled. There are a couple different layers to this:

1. Key "handlers", which detect monogame input events and call the appropriate controller methods. This essentially binds a key input to some meaningful game action.
2. Key "controllers", which orchestrate the logic associated with a meaningful game action.
3. Input "validators", which determine whether it's ok for the meaningful game action to take place. This is essentially a factoring out of logic that would normally reside in a controller.

I don't see this logic needing to be repeated in all three places. IMO the way to approach this would be to put the logic in the validators, and handle the validation result appropriately (status message or no-op depending on what the vanilla client does). I don't think it makes sense in the handlers, because those are concerned only with detecting monogame events, and I don't think it makes sense in the controllers because other validations are already encapsulated by the validators.